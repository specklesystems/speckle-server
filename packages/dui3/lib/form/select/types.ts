export type ProjectsSelectItemType = { id?: string; name: string }
export type ModelsSelectItemType = { id?: string; name: string }
export type VersionsSelectItemType = {
  id: string
  message: string
  createdAt: Date
  referencedObject: string
  sourceApplication: string
}
