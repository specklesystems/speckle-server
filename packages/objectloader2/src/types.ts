export type CustomLogger = (message?: string, ...optionalParams: unknown[]) => void

export interface Item {
  id: string
  obj: Base
}

export interface Base {
  id: string
  __closure?: string[]
}

export function isString(value: unknown): value is string {
  return typeof value === 'string'
}
