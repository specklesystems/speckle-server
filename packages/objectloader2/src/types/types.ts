export type CustomLogger = (message?: string, ...optionalParams: unknown[]) => void

export type Fetcher = (
  input: string | URL | Request,
  init?: RequestInit
) => Promise<Response>

export interface Item {
  baseId: string
  base: Base
  size?: number
}

export interface Base {
  id: string
  speckle_type: string
  __closure?: Record<string, number>
}

export interface Reference {
  speckle_type: string
  referencedId: string
  __closure?: Record<string, number>
}

export interface DataChunk extends Base {
  data?: Base[]
}

export function isBase(maybeBase?: unknown): maybeBase is Base {
  return (
    maybeBase !== null &&
    typeof maybeBase === 'object' &&
    'id' in maybeBase &&
    typeof maybeBase.id === 'string'
  )
}

export function isReference(maybeRef?: unknown): maybeRef is Reference {
  return (
    maybeRef !== null &&
    typeof maybeRef === 'object' &&
    'referencedId' in maybeRef &&
    typeof maybeRef.referencedId === 'string'
  )
}

export function isScalar(
  value: unknown
): value is string | number | boolean | bigint | symbol | undefined {
  const type = typeof value
  return (
    value === null ||
    type === 'string' ||
    type === 'number' ||
    type === 'boolean' ||
    type === 'bigint' ||
    type === 'symbol' ||
    type === 'undefined'
  )
}

export function take<T>(it: Iterator<T>, count: number): T[] {
  const result: T[] = []
  for (let i = 0; i < count; i++) {
    const itr = it.next()
    if (itr.done) break
    result.push(itr.value)
  }
  return result
}
