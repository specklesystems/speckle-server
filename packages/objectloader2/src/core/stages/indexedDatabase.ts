/* eslint-disable @typescript-eslint/no-unsafe-function-type */
import { Item } from '../../types/types.js'
import { Database } from '../interfaces.js'
import BatchingQueue from '../../queues/batchingQueue.js'
import { ItemStore } from './ItemStore.js'

export interface IndexedDatabaseOptions {
  indexedDB?: IDBFactory
  keyRange?: {
    bound: Function
    lowerBound: Function
    upperBound: Function
  }
}

export default class IndexedDatabase implements Database {
  #options: IndexedDatabaseOptions
  #cacheDB: ItemStore
  #writeQueue: BatchingQueue<Item> | undefined

  constructor(options: IndexedDatabaseOptions) {
    this.#options = options
    this.#cacheDB = new ItemStore(
      {
        indexedDB: this.#options.indexedDB,
        keyRange: this.#options.keyRange
      },
      'speckle-cache',
      'cache'
    )
  }

  async getAll(keys: string[]): Promise<(Item | undefined)[]> {
    await this.#cacheDB.init()
    return await this.#cacheDB.bulkGet(keys)
  }

  async saveBatch(params: { batch: Item[] }): Promise<void> {
    await this.#cacheDB.init()
    const { batch } = params
    await this.#cacheDB.bulkInsert(batch)
  }

  async disposeAsync(): Promise<void> {
    await this.#writeQueue?.disposeAsync()
    this.#writeQueue = undefined
    this.#cacheDB.close()
  }
}
