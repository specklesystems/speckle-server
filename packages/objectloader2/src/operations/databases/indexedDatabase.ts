/* eslint-disable @typescript-eslint/no-unsafe-function-type */
import BatchingQueue from '../../helpers/batchingQueue.js'
import { CustomLogger, Item } from '../../types/types.js'
import { isSafari } from '@speckle/shared'
import { Dexie, DexieOptions, Table } from 'dexie'
import { Database } from '../interfaces.js'

class ObjectStore extends Dexie {
  static #databaseName: string = 'speckle-cache'
  objects!: Table<Item, string> // Table type: <entity, primaryKey>

  constructor(options: DexieOptions) {
    super(ObjectStore.#databaseName, options)

    this.version(1).stores({
      objects: 'baseId, item' // baseId is primary key
    })
  }
}

export interface IndexedDatabaseOptions {
  logger?: CustomLogger
  indexedDB?: IDBFactory
  keyRange?: {
    bound: Function
    lowerBound: Function
    upperBound: Function
  }
}

export default class IndexedDatabase implements Database {
  #options: IndexedDatabaseOptions
  #logger: CustomLogger

  #cacheDB?: ObjectStore

  #writeQueue: BatchingQueue<Item> | undefined

  // #count: number = 0

  constructor(options: IndexedDatabaseOptions) {
    this.#options = options
    this.#logger = options.logger || ((): void => {})
  }

  async getAll(keys: string[]): Promise<(Item | undefined)[]> {
    await this.#setupCacheDb()
    let items: (Item | undefined)[] = []
    //   this.#count++
    //  const startTime = performance.now()
    // this.#logger('Start read ' + x + ' ' + batch.length)

    //faster than BulkGet with dexie
    await this.#cacheDB!.transaction('r', this.#cacheDB!.objects, async () => {
      const gets = keys.map((key) => this.#cacheDB!.objects.get(key))
      const cachedData = await Promise.all(gets)
      items = cachedData
    })
    // const endTime = performance.now()
    // const duration = endTime - startTime
    //this.#logger('Saved batch ' + x + ' ' + batch.length + ' ' + duration / TIME_MS.second)

    return items
  }

  async #openDatabase(): Promise<ObjectStore> {
    const db = new ObjectStore({
      indexedDB: this.#options.indexedDB ?? globalThis.indexedDB,
      IDBKeyRange: this.#options.keyRange ?? IDBKeyRange,
      chromeTransactionDurability: 'relaxed'
    })
    await db.open()
    return db
  }

  async #setupCacheDb(): Promise<void> {
    if (this.#cacheDB !== undefined) {
      return
    }

    // Initialize
    await this.#safariFix()
    this.#cacheDB = await this.#openDatabase()
  }

  //this is for testing only - in the real world we will not use this
  async add(item: Item): Promise<void> {
    await this.#setupCacheDb()
    await this.#cacheDB!.transaction('rw', this.#cacheDB!.objects, async () => {
      return await this.#cacheDB?.objects.add(item)
    })
  }

  async getItem(params: { id: string }): Promise<Item | undefined> {
    const { id } = params
    await this.#setupCacheDb()
    //might not be in the real DB yet, so check the write queue first
    if (this.#writeQueue) {
      const item = this.#writeQueue.get(id)
      if (item) {
        return item
      }
    }

    return this.#cacheDB!.transaction('r', this.#cacheDB!.objects, async () => {
      return await this.#cacheDB?.objects.get(id)
    })
  }

  async cacheSaveBatch(params: { batch: Item[] }): Promise<void> {
    await this.#setupCacheDb()
    const { batch } = params
    //const x = this.#count
    //this.#count++

    // const startTime = performance.now()
    //  this.#logger('Start save ' + x + ' ' + batch.length)
    await this.#cacheDB!.objects.bulkPut(batch)
    // const endTime = performance.now()
    // const duration = endTime - startTime
    //this.#logger('Saved batch ' + x + ' ' + batch.length + ' ' + duration / TIME_MS.second)
  }

  /**
   * Fixes a Safari bug where IndexedDB requests get lost and never resolve - invoke before you use IndexedDB
   * @link Credits and more info: https://github.com/jakearchibald/safari-14-idb-fix
   */
  async #safariFix(): Promise<void> {
    // No point putting other browsers or older versions of Safari through this mess.
    if (!isSafari() || !this.#options.indexedDB?.databases) return Promise.resolve()

    let intervalId: ReturnType<typeof setInterval>

    return new Promise<void>((resolve: () => void) => {
      const tryIdb = (): Promise<IDBDatabaseInfo[]> | undefined =>
        this.#options.indexedDB?.databases().finally(resolve)
      intervalId = setInterval(() => {
        void tryIdb()
      }, 100)
      void tryIdb()
    }).finally(() => clearInterval(intervalId))
  }

  async disposeAsync(): Promise<void> {
    this.#cacheDB?.close()
    this.#cacheDB = undefined
    await this.#writeQueue?.disposeAsync()
    this.#writeQueue = undefined
  }
}
