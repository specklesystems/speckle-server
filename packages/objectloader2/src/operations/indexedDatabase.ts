import BatchingQueue from '../helpers/batchingQueue.js'
import Queue from '../helpers/queue.js'
import { CustomLogger, Item } from '../types/types.js'
import { isSafari } from '@speckle/shared'
import { BaseDatabaseOptions } from './options.js'
import { Cache } from './interfaces.js'
import { Dexie, DexieOptions, Table } from 'dexie'

class ObjectStore extends Dexie {
  static #databaseName: string = 'speckle-cache'
  objects!: Table<Item, string> // Table type: <entity, primaryKey>

  constructor(options: DexieOptions) {
    super(ObjectStore.#databaseName, options)

    this.version(1).stores({
      objects: 'baseId' // baseId is primary key
    })
  }
}

export default class IndexedDatabase implements Cache {
  #options: BaseDatabaseOptions
  #logger: CustomLogger

  #cacheDB?: ObjectStore

  #writeQueue: BatchingQueue | undefined

  // #count: number = 0

  constructor(options: BaseDatabaseOptions) {
    this.#options = {
      ...{
        maxCacheReadSize: 10000,
        maxCacheBatchWriteWait: 1000
      },
      ...options
    }
    this.#logger = options.logger || (() => {})
  }

  async add(item: Item): Promise<void> {
    if (!this.#writeQueue) {
      await this.#setupCacheDb()
      this.#writeQueue = new BatchingQueue({
        batchSize: this.#options.maxCacheWriteSize ?? 10000,
        maxWaitTime: this.#options.maxCacheBatchWriteWait,
        processFunction: (batch: Item[]) =>
          this.#cacheSaveBatch({ batch, cacheDB: this.#cacheDB! })
      })
    }
    this.#writeQueue.add(item)
  }

  async disposeAsync(): Promise<void> {
    await this.#writeQueue?.disposeAsync()
  }

  async #openDatabase(): Promise<ObjectStore> {
    const db = new ObjectStore({
      indexedDB: this.#options.indexedDB ?? globalThis.indexedDB,
      IDBKeyRange: this.#options.keyRange ?? IDBKeyRange,
      chromeTransactionDurability: 'relaxed'
    })
    await db.open()
    return db
  }

  async #setupCacheDb(): Promise<void> {
    if (this.#cacheDB !== undefined) {
      return
    }

    // Initialize
    await this.#safariFix()
    this.#cacheDB = await this.#openDatabase()
  }

  async processItems(params: {
    ids: string[]
    foundItems: Queue<Item>
    notFoundItems: Queue<string>
  }): Promise<void> {
    const { ids, foundItems, notFoundItems } = params
    await this.#setupCacheDb()
    const maxCacheReadSize = this.#options.maxCacheReadSize ?? 10000

    for (let i = 0; i < ids.length; ) {
      if ((this.#writeQueue?.count() ?? 0) > maxCacheReadSize * 2) {
        this.#logger(
          'pausing reads (# in write queue: ' + this.#writeQueue?.count() + ')'
        )
        await new Promise((resolve) => setTimeout(resolve, 1000)) // Pause for 1 second, protects against out of memory
        continue
      }
      const batch = ids.slice(i, i + maxCacheReadSize)
      //   const x = this.#count
      //   this.#count++
      //  const startTime = performance.now()
      // this.#logger('Start read ' + x + ' ' + batch.length)

      //faster than BulkGet with dexie
      await this.#cacheDB!.transaction('r', this.#cacheDB!.objects, async () => {
        const gets = batch.map((key) => this.#cacheDB!.objects.get(key))
        const cachedData = await Promise.all(gets)
        for (let i = 0; i < cachedData.length; i++) {
          if (cachedData[i]) {
            foundItems.add(cachedData[i]!)
          } else {
            notFoundItems.add(batch[i])
          }
        }
      })
      // const endTime = performance.now()
      // const duration = endTime - startTime
      // this.#logger('Read batch ' + x + ' ' + batch.length + ' ' + duration / 1000)

      // interate down here to help with pausing
      i += maxCacheReadSize
    }
  }

  async getItem(params: { id: string }): Promise<Item | undefined> {
    const { id } = params
    await this.#setupCacheDb()
    const itemToBeWritten = this.#writeQueue?.get(id)
    if (itemToBeWritten) {
      return itemToBeWritten
    }
    return this.#cacheDB!.transaction('r', this.#cacheDB!.objects, async () => {
      return await this.#cacheDB?.objects.get(id)
    })
  }

  async #cacheSaveBatch(params: {
    batch: Item[]
    cacheDB: ObjectStore
  }): Promise<void> {
    const { batch, cacheDB } = params
    //const x = this.#count
    //this.#count++

    // const startTime = performance.now()
    //  this.#logger('Start save ' + x + ' ' + batch.length)
    await cacheDB.objects.bulkPut(batch)
    // const endTime = performance.now()
    // const duration = endTime - startTime
    //this.#logger('Saved batch ' + x + ' ' + batch.length + ' ' + duration / 1000)
  }

  /**
   * Fixes a Safari bug where IndexedDB requests get lost and never resolve - invoke before you use IndexedDB
   * @link Credits and more info: https://github.com/jakearchibald/safari-14-idb-fix
   */
  async #safariFix(): Promise<void> {
    // No point putting other browsers or older versions of Safari through this mess.
    if (!isSafari() || !this.#options.indexedDB?.databases) return Promise.resolve()

    let intervalId: ReturnType<typeof setInterval>

    return new Promise<void>((resolve: () => void) => {
      const tryIdb = () => this.#options.indexedDB?.databases().finally(resolve)
      intervalId = setInterval(() => {
        void tryIdb()
      }, 100)
      void tryIdb()
    }).finally(() => clearInterval(intervalId))
  }
}
