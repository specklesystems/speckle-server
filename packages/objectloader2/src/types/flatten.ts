import { Base } from './types.js'

/**
 * Flattens a speckle object. It will ignore arrays, null and undefined valuesm, as well as various 'safe to ignore' speckle properties, such as
 * bbox, __closure, __parents, totalChildrenCount.
 * @param obj object to flatten
 * @returns an object with all its props flattened into `prop.subprop.subsubprop`.
 */
export function flattenBase(obj: Base): Record<string, string | number> {
  const flatten = {} as Record<string, string | number>
  for (const k in obj) {
    if (
      k === 'id' ||
      k === '__closure' ||
      k === '__parents' ||
      k === 'bbox' ||
      k === 'totalChildrenCount'
    ) {
      continue
    }

    const v = obj[k]
    if (v === null || v === undefined || Array.isArray(v)) {
      continue
    }
    if (v.constructor === Object) {
      const flattenProp = flattenBase(v as unknown as Base)
      for (const pk in flattenProp) {
        flatten[k + '.' + pk] = flattenProp[pk]
      }
      continue
    }
    const type = typeof v
    if (type === 'string' || type === 'number') {
      flatten[k] = v as string | number
    } else if (type === 'boolean') {
      flatten[k] = v ? 'true' : 'false' // Convert boolean to string for consistency
    }
  }
  if (obj.id) {
    flatten.id = obj.id
  }
  return flatten
}

export function sententizeBase(obj: Base): string {
  const flattened = flattenBase(obj)
  const propertyStrings = Object.entries(flattened).map(([key, value]) => {
    // Make the key more readable (e.g., "Fire Rating" instead of "FireRating")
    const formattedKey = key.replace(/([A-Z])/g, ' $1').trim()
    return `${formattedKey} is ${value}`
  })
  // Join the individual strings with a comma and a space.
  return `This object has the following properties: ${propertyStrings.join(', ')}.`
}
