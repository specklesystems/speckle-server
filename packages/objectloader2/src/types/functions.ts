import { Base, Reference } from './types.js'

export type CustomLogger = (message?: string, ...optionalParams: unknown[]) => void

export type Fetcher = (
  input: string | URL | Request,
  init?: RequestInit
) => Promise<Response>

export function isBase(maybeBase?: unknown): maybeBase is Base {
  return (
    maybeBase !== null &&
    typeof maybeBase === 'object' &&
    'id' in maybeBase &&
    typeof maybeBase.id === 'string'
  )
}

export function isReference(maybeRef?: unknown): maybeRef is Reference {
  return (
    maybeRef !== null &&
    typeof maybeRef === 'object' &&
    'referencedId' in maybeRef &&
    typeof maybeRef.referencedId === 'string'
  )
}

export function isScalar(
  value: unknown
): value is string | number | boolean | bigint | symbol | undefined {
  const type = typeof value
  return (
    value === null ||
    type === 'string' ||
    type === 'number' ||
    type === 'boolean' ||
    type === 'bigint' ||
    type === 'symbol' ||
    type === 'undefined'
  )
}

export function take<T>(it: Iterator<T>, count: number): T[] {
  const result: T[] = []
  for (let i = 0; i < count; i++) {
    const itr = it.next()
    if (itr.done) break
    result.push(itr.value)
  }
  return result
}

export enum ObjectLoader2Flags {
  DEBUG = 'debug',
  USE_CACHE = 'useCache'
}

const defaultValues: Record<ObjectLoader2Flags, string> = {
  [ObjectLoader2Flags.DEBUG]: 'false',
  [ObjectLoader2Flags.USE_CACHE]: 'true'
}

export function getFeatureFlag(
  paramName: ObjectLoader2Flags,
  useDefault: boolean = true
): string | undefined {
  // Check if the code is running in a browser environment üåê
  const isBrowser =
    typeof window !== 'undefined' && typeof window.document !== 'undefined'

  if (!isBrowser) {
    // If in Node.js or another server environment, return the default
    return useDefault ? defaultValues[paramName] : undefined
  }

  // In a browser, parse the query string
  const params = new URLSearchParams(window.location.search)

  // .get() returns the value, or null if it's not found.
  // The nullish coalescing operator (??) provides the default value
  // if the left-hand side is null or undefined.
  return params.get(paramName) ?? (useDefault ? defaultValues[paramName] : undefined)
}

/**
 * Finds the first index of a "needle" Uint8Array within a "haystack" Uint8Array.
 * @param haystack The larger array to search within.
 * @param needle The smaller array to search for.
 * @param start The index to start searching from. Defaults to 0.
 * @returns The starting index of the needle, or -1 if not found.
 */
export function indexOf(
  haystack: Uint8Array,
  needle: Uint8Array,
  start: number = 0
): number {
  if (needle.length === 0) {
    return 0
  }

  // The last possible starting position for a match
  const limit = haystack.length - needle.length

  for (let i = start; i <= limit; i++) {
    let foundMatch = true
    for (let j = 0; j < needle.length; j++) {
      if (haystack[i + j] !== needle[j]) {
        foundMatch = false
        break // Mismatch, break inner loop
      }
    }
    if (foundMatch) {
      return i // Found a full match at index i
    }
  }

  return -1 // No match found
}
