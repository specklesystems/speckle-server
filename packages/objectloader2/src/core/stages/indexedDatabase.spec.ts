import { describe, it, expect, beforeEach, afterEach } from 'vitest'
import { IDBFactory, IDBKeyRange } from 'fake-indexeddb'

import { Base, Item } from '../../types/types.js'
import {IndexedDatabase, IndexedDatabaseOptions } from './indexedDatabase.js'

// Mock Item
const defaultItem = (id: string): Item => ({
  baseId: id,
  base: { foo: 'bar' } as unknown as Base
})

describe('IndexedDatabase', () => {
  let db: IndexedDatabase
  let options: IndexedDatabaseOptions

  beforeEach(() => {
    options = { indexedDB: new IDBFactory(), keyRange: IDBKeyRange }
    db = new IndexedDatabase(options)
  })

  afterEach(() => {
    db.dispose()
  })

  it('should add and get multiple items', async () => {
    const items = [defaultItem('id1'), defaultItem('id2')]
    await db.putAll(items)
    const result = await db.getAll(['id1', 'id2'])
    expect(result).toMatchSnapshot()
    expect(result).toEqual(items)
  })

  it('should dispose without error', () => {
    expect(() => db.dispose()).not.toThrow()
  })
})
