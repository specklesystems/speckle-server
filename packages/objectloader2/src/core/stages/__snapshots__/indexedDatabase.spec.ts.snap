// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`IndexedDatabase > should add and get multiple items 1`] = `
[
  {
    "base": {
      "foo": "bar",
    },
    "baseId": "id1",
  },
  {
    "base": {
      "foo": "bar",
    },
    "baseId": "id2",
  },
]
`;
